### PR TITLE
Updated pytorch integration

### PR DESCRIPTION
**Pull request by jakubczakon**
* improved pytorch integration
* fixed colab/github/neptune linked buttons

### CODE REVIEW

File: docs/integrations/pytorch.rst [hunk 1]
```diff
@@ -3,109 +3,370 @@
...
+
+    - :ref:`PyTorch Lightning <integrations-pytorch-lightning>`
+    - :ref:`Fastai and Fastai2 <integrations-fastai>`
+    - :ref:`PyTorch Ignite <integrations-pytorch-ignite>`
+    - :ref:`Catalyst <integrations-catalyst>`
+    - :ref:`Skorch <integrations-skorch>`
+
+    For pure PyTorch integration, read on.
+
+|colab-script-neptune|
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:19]: When I land on example project, I see empty'ish dashboard.

File: docs/integrations/pytorch.rst [hunk 2]
```diff
@@ -3,109 +3,370 @@
...
+
+What will you get with this integration?
+----------------------------------------
+
+|pytorch-tour-loom|
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:24]: it links to optuna video

File: docs/integrations/pytorch.rst [hunk 3]
```diff
@@ -3,109 +3,370 @@
...
+
+|Pytorch| is an open source machine learning framework commonly used for building deep neural network models.
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:26]: machine learning -> maybe deep learning

File: docs/integrations/pytorch.rst [hunk 4]
```diff
@@ -3,109 +3,370 @@
...
+
+With Neptune + PyTorch integration you can:
+
+- log hyperparameters
+- see learning curves for losses and metrics during training
+- see hardware consumption during training
+- log stdout and stderr
+- log training code and .git information
+- log model weights
+- log torch tensors as images
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:37]: this is cool, I would put higher

File: docs/integrations/pytorch.rst [hunk 5]
```diff
@@ -3,109 +3,370 @@
...
+
+|colab-script-neptune|
+
+.. _pytorch-quickstart:
+
+Quickstart
+----------
+This quickstart will show you how to:
+
+* Install the necessary neptune packages
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:62]: package

File: docs/integrations/pytorch.rst [hunk 6]
```diff
@@ -3,109 +3,370 @@
...
+* Connect Neptune to your Optuna hyperparameter tuning code and create the first experiment
+* Log metrics, figures, and artifacts from your first Optuna sweep to Neptune, and
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:64]: this point and one above is about optuna

File: docs/integrations/pytorch.rst [hunk 7]
```diff
@@ -3,109 +3,370 @@
...
+You have ``Python 3.x`` and following libraries installed:
+
+* ``neptune-client``. See :ref:`neptune-client installation guide <installation-neptune-client>`.
+* ``torch``. See |pytorch-install|.
+
+You also need minimal familiarity with torch. Have a look at this |pytorch-guide| to get started.
+
+.. code-block:: bash
+
+   pip install --quiet torch neptune-client
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:80]: I would not add it here. linking to torch docs is good enough

File: docs/integrations/pytorch.rst [hunk 8]
```diff
@@ -3,109 +3,370 @@
...
+
+    import neptune
+
+    neptune.init(api_token='ANONYMOUS', project_qualified_name='shared/pytorch-integration')
+
+.. tip::
+
+    You can also use your personal API token. Read more about how to :ref:`securely set the Neptune API token <how-to-setup-api-token>`.
+
+Step 2: Create an Experiment and log parameters
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:96]: you did not log params in this step.

Perhaps you can cover init and create_exp in the step 1.

File: docs/integrations/pytorch.rst [hunk 9]
```diff
@@ -3,109 +3,370 @@
...
+.. code-block:: python3
+
+    neptune.create_experiment('pytorch-quickstart')
+
+This also creates a link to the experiment. Open the link in a new tab.
+The charts will currently be empty, but keep the window open. You will be able to see live metrics once logging starts.
+
+Step 3: Add logging into your training loop
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Log your loss after every batch by adding ``neptune.log_metric`` inside of the loop.
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:109]: maybe:
inside of the training loop.

File: docs/integrations/pytorch.rst [hunk 10]
```diff
@@ -3,109 +3,370 @@
...
+---------------
+To get started with this integration, follow the :ref:`quickstart <pytorch-quickstart>` below.
+You can also skip the basics and take a look at how to log model weights and prediction images in the :ref:`advanced options <pytorch-advanced-options>` section.
 
-Log PyTorch metrics
--------------------
-It is very easy to track PyTorch metrics in Neptune. The following example presents a simple integration that lets you track loss:
+If you want to try things out and focus only on the code you can either:
+
+|colab-script-neptune|
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:54]: this trio is really nice!

File: docs/integrations/pytorch.rst [hunk 11]
```diff
@@ -3,109 +3,370 @@
 Neptune-PyTorch Integration
 ===========================
 
-Neptune has implemented an integration with the PyTorch machine learning library.
+.. note::
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:6]: Like it as an idea. However, placing it at the very top suggest that it is most important - it looks like info banner.

If I landed here to learn about PyTorch, then it is not that crucial for me.

Unless, you did it to have half of the video visible at the first glance. If so, then cool :)

File: docs/integrations/pytorch.rst [hunk 12]
```diff
@@ -3,109 +3,370 @@
...
+        if batch_idx == 100:
+            break
 
-.. image:: ../_static/images/integrations/ht-log-pytorch-1.png
-   :target: ../_static/images/integrations/ht-log-pytorch-1.png
-   :alt: image
+You can log epoch metric and losses by calling ``neptune.log_metric`` at the epoch level.
 
-.. note::
+Step 4: Run your training script
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:128]: Here I'm a bit confused.
I was not asked to create any script - I just saw snippet about where to place log_metric.

Also, in the step one you suggest to run the code. Where should I run it?

File: docs/integrations/pytorch.rst [hunk 13]
```diff
@@ -3,109 +3,370 @@
...
+----------------
+
+Log hardware consumption
+^^^^^^^^^^^^^^^^^^^^^^^^
+Neptune can automatically log your CPU and GPU consumption during training as well as stderr and stdout from your console.
+To do that you just need to install psutil.
+
+.. code-block:: bash
+
+    pip install psutil
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:158]: Maybe link to their docs?

File: docs/integrations/pytorch.rst [hunk 14]
```diff
@@ -3,109 +3,370 @@
...
-                       train=True,
-                       download=True,
-                       transform=transforms.Compose([transforms.ToTensor(),
-                                                     transforms.Normalize((0.1307,), (0.3081,))])),
-       batch_size=PARAMS['batch_size'],
-       shuffle=True)
-
-    model = Net().to(PARAMS['DEVICE'])
+    PARAMS = {'lr':0.005,
+               'momentum':0.9,
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:174]: Perhaps we want this dict to be aligned vertically.

File: docs/integrations/pytorch.rst [hunk 15]
```diff
@@ -3,109 +3,370 @@
...
+               'iterations':100}
 
     optimizer = optim.SGD(model.parameters(), PARAMS['lr'], PARAMS['momentum'])
 
+    # log params
+    neptune.create_experiment('pytorch-advanced', params=PARAMS)
+
+.. image:: ../_static/images/integrations/pytorch-parameters.png
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:182]: Bottom half of the screenshot is not very useful here.

File: docs/integrations/pytorch.rst [hunk 16]
```diff
@@ -3,109 +3,370 @@
...
+   :target: ../_static/images/integrations/pytorch-parameters.png
+   :alt: PyTorch hyperparameter logging
+
+|example-hyperparameters|
+
+Log model weights
+^^^^^^^^^^^^^^^^^
+You can log model weights to Neptune both during and after training.
+
+To do that just use a ``neptune.log_artifact`` method on the saved model file.
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:192]: If you do `:meth:method` not ``method``, then users can click on it as see docstrings.

File: docs/integrations/pytorch.rst [hunk 17]
```diff
@@ -3,109 +3,370 @@
...
 
+.. image:: ../_static/images/integrations/pytorch-images.png
+   :target: ../_static/images/integrations/pytorch-images.png
+   :alt: PyTorch logging images
+
+|example-images|
+
+Remember that you can try it out with zero setup:
+
+|colab-script-neptune|
```
Comment (kamil-kaczmarek) [inline docs/integrations/pytorch.rst:245]: I like this repetition

**Review by kamil-kaczmarek — APPROVED** [review]
I left some comments, overall nothing is blocking from my side.

File: docs/integrations/pytorch.rst (hunk 11 above)
Comment (jakubczakon) [reply docs/integrations/pytorch.rst:6]: Yeah, I do think that showing those higher-level libs (where you get more value with the integration) is very important. 

MLflow talks about integration with pytorch with explicit dependency of pytorch-lightning even.

File: docs/integrations/pytorch.rst (hunk 2 above)
Comment (jakubczakon) [reply docs/integrations/pytorch.rst:24]: yep, no video ready yet -> waiting for some neptune-client merges

File: docs/integrations/pytorch.rst (hunk 3 above)
Comment (jakubczakon) [reply docs/integrations/pytorch.rst]: took it from their site actually, but I think you are right -> it is in deep learning frameworks in our docs right?

**Pull request merged**